docs/design/sum_types.md

```diff
@@ -0,0 +1,200 @@
+have limited flexibility. There is no way to control the representation of a
+`choice` type, or define methods or other members for it (although you can
+extend it to implement interfaces, using
+[`external impl`](generics/overview.md#implementing-interfaces) or
+[`adapter`](generics/overview.md#adapting-types)). However, a `class` type can
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants, and the internal storage
+layout will presumably involve untagged unions or some other low-level storage
+primitive which hasn't been designed yet, but the key to defining a sum type's
+interface is enabling it to support pattern matching. To do that, the sum type
+has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+Here's how that would look if `Optional` were defined as a class:
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  let None:! Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some[addr me: Self*](value: T) -> ReturnType;
+    fn None[addr me: Self*]() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
+    fn Op[me: Self, Continuation:! MatchContinuation](
+        continuation: Continuation*) -> Continuation.ReturnType {
+      if (me.has_value) {
+        return continuation->Some(me.value);
+      } else {
+        return continuation->None();
+      }
+    }
+  }
+
+  // Operations like destruction, copying, assignment, and comparison are
+  // omitted for brevity.
+}
+```
+
+In this code, `Optional` makes itself available for use in pattern matching by
+declaring that it implements the `Match` interface. `Match` takes an interface
+argument, called the _continuation interface_, which specifies the set of
+possible alternatives by declaring a method for each one. In this case, we pass
+`Optional.MatchContinuation` as the continuation interface.
+
+When compiling a `match` statement, the compiler checks that the type being
+matched implements `Match(C)` for some continuation interface `C`. Then, it
+notionally transforms the `match` body into a class that implements `C`, with
+one method for each `case`, and then passes that to `Match.Op` on the object
+being matched. For example, the `match` statement shown earlier might be
```

Comment:
What pattern syntax does this apply to? Only `.Name` or `.Name(args)`? Or does this also apply to `Optional(T).Name` and `Optional(T).Name(args)` (presumably looking for `Optional(T) as Match` regardless of the type of the scrutinee)? Does this apply only to proper patterns or also to expression patterns?

The decision here might motivate having a specific introducer syntax for expression patterns.